File: solutions/python/resistor-color-trio/1/resistor_color_trio.py
COLOUR_MAP = {
    "black": 0,
    "brown": 1,
    "red": 2,
    "orange": 3,
    "yellow": 4,
    "green": 5,
    "blue": 6,
    "violet": 7,
    "grey": 8,
    "white": 9,
}

def value(colors):
    return 10 * COLOUR_MAP[colors[0]] + COLOUR_MAP[colors[1]]

def label(colors):
    main_value = value(colors[:2])
    multiplier = 10 ** COLOUR_MAP[colors[2]]
    resistance = main_value * multiplier

    if resistance >= 10**9:
        return f"{resistance // 10**9} gigaohms"
    elif resistance >= 10**6:
        return f"{resistance // 10**6} megaohms"
    elif resistance >= 10**3:
        return f"{resistance // 10**3} kiloohms"
    else:
        return f"{resistance} ohms"

File: elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @color_codes %{
    :black => 0,
    :brown => 1,
    :red => 2,
    :orange => 3,
    :yellow => 4,
    :green => 5,
    :blue => 6,
    :violet => 7,
    :grey => 8,
    :white => 9,
  }

  @metric [:ohms, :kiloohms, :megaohms, :gigaohms]

  def add_label(number, unit) do
    if number > 999 do
      add_label(number / 1000, unit+1)
    else
      {number, Enum.at(@metric, unit)}
    end
  end

  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label([tens, units, magnitude | _]) do
    add_label((10*@color_codes[tens] + @color_codes[units]) * 10**@color_codes[magnitude], 0)
  end
end
